Source/DafnyLanguageServer/Workspace/IdeState.cs

```diff
@@ -426,104 +428,107 @@ public record IdeState(
+    var view = new IdeVerificationTaskState(range, PublishedVerificationStatus.Error, diagnostics.ToList(), hitErrorLimit, boogieException.Task, new Stale());
 
     return previousState with {
-      VerificationResults = previousState.VerificationResults.SetItem(uri,
-        previousState.VerificationResults[uri].SetItem(range, previousVerificationResult with {
-          Implementations = previousVerificationResult.Implementations.SetItem(name, view)
+      CanVerifyStates = previousState.CanVerifyStates.SetItem(uri,
+        previousState.CanVerifyStates[uri].SetItem(range, previousVerificationResult with {
+          VerificationTasks = previousVerificationResult.VerificationTasks.SetItem(name, view)
         }))
     };
   }
 
-  private IdeState UpdateBoogieUpdate(DafnyOptions options, ILogger logger, BoogieUpdate boogieUpdate) {
+  private IdeState HandleBoogieUpdate(DafnyOptions options, ILogger logger, BoogieUpdate boogieUpdate) {
     var previousState = this;
-    UpdateGutterIconTrees(boogieUpdate, options, logger);
 
-    var name = Compilation.GetImplementationName(boogieUpdate.ImplementationTask.Implementation);
+    var name = Compilation.GetTaskName(boogieUpdate.VerificationTask);
     var status = StatusFromBoogieStatus(boogieUpdate.BoogieStatus);
     var uri = boogieUpdate.CanVerify.Tok.Uri;
     var range = boogieUpdate.CanVerify.NameToken.GetLspRange();
 
-    var previousVerificationResult = previousState.VerificationResults[uri][range];
-    var previousImplementations = previousVerificationResult.Implementations;
-    var previousView = previousImplementations.GetValueOrDefault(name) ??
-                       new IdeImplementationView(range, status, Array.Empty<Diagnostic>(), false);
+    var previousVerificationResult = previousState.CanVerifyStates[uri][range];
+    var previousImplementations = previousVerificationResult.VerificationTasks;
+    var previousView = previousImplementations.GetValueOrDefault(name);
     var counterExamples = previousState.Counterexamples;
-    bool hitErrorLimit = previousView.HitErrorLimit;
-    var diagnostics = previousView.Diagnostics;
+    bool hitErrorLimit = previousView?.HitErrorLimit ?? false;
+    IVerificationStatus rawStatus = boogieUpdate.BoogieStatus;
+    var diagnostics = previousView?.Diagnostics ?? Array.Empty<Diagnostic>();
     if (boogieUpdate.BoogieStatus is Running) {
       diagnostics = Array.Empty<Diagnostic>();
       counterExamples = Array.Empty<Counterexample>();
       hitErrorLimit = false;
     }
 
-    if (boogieUpdate.BoogieStatus is BatchCompleted batchCompleted) {
-      counterExamples = counterExamples.Concat(batchCompleted.VcResult.counterExamples);
-      hitErrorLimit |= batchCompleted.VcResult.maxCounterExamples == batchCompleted.VcResult.counterExamples.Count;
+    if (boogieUpdate.BoogieStatus is Completed completed) {
+      // WarnContradictoryAssumptions should be computable after completing a single assertion batch.
```

Comment:
This will require some refactoring, though I think it shouldn't be too hard. The key issue is that, at the moment, we look for all proof obligations associated with a given definition and report any that don't show up in the set of covered elements. To make it work per assertion batch, we need to update that logic to identify proof obligations associated with that assertion batch, and see which ones are missing.